File: Data Structures & Algorithms/search-2d-matrix/submission-2.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {

        int m = matrix.length;
        int n = matrix[0].length;

        int low = 0;
        int high = m * n - 1;

        while (low <= high) {
            int middlePosition = low + (high - low) / 2;
            int value = matrix[middlePosition / n][middlePosition % n];

            if (target == value) {
                return true;
            } else if (target > value) {
                low = middlePosition + 1;
            } else {
                high = middlePosition - 1;
            }
        }

        return false;
    }

}

File: Data Structures & Algorithms/binary-search/submission-2.ts
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number}
     */
    search(nums: number[], target: number): number {
        let low = 0;
        let high = nums.length - 1;

        while (low <= high) {
            console.log("----Start-----");
            console.log("low = " + low);
            console.log("high = " + high);
            const middlePosition = low + Math.floor((high - low) / 2);
            const valueAtMiddlePosition = nums[middlePosition];
            console.log("middlePosition = " + middlePosition);
            console.log("valueAtMiddlePosition = " + valueAtMiddlePosition);

            if (target === valueAtMiddlePosition) {
                return middlePosition;
            } else if (target < valueAtMiddlePosition) {
                high = middlePosition - 1;
            } else {
                low = middlePosition + 1;
            }
        }

        return -1;
    }
}
